File: OpenHome/Net/Bindings/Cs/Device/Tests/TestBasicDv.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading;
using System.Diagnostics;
using OpenHome.Net.Device;
using OpenHome.Net.Device.Providers;

namespace OpenHome.Net
{
    class ProviderTestBasic : DvProviderOpenhomeOrgTestBasic1
    {
        public ProviderTestBasic(DvDevice aDevice)
            : base(aDevice)
        {
            EnablePropertyVarUint();
            EnablePropertyVarInt();
            EnablePropertyVarBool();
            EnablePropertyVarStr();
            EnablePropertyVarBin();
            SetPropertyVarUint(0);
            SetPropertyVarInt(0);
            SetPropertyVarBool(false);
            string empty = "";
            byte[] emptyBinary = new byte[0];
            SetPropertyVarStr(empty);
            SetPropertyVarBin(emptyBinary);

            EnableActionIncrement();
            EnableActionDecrement();
            EnableActionToggle();
            EnableActionEchoString();
            EnableActionEchoBinary();
            EnableActionSetUint();
            EnableActionGetUint();
            EnableActionSetInt();
            EnableActionGetInt();
            EnableActionSetBool();
            EnableActionGetBool();
            EnableActionSetString();
            EnableActionGetString();
            EnableActionSetBinary();
            EnableActionGetBinary();
            EnableActionSetMultiple();
        }

        protected override void Increment(IDvInvocation aInvocation, uint aValue, out uint aResult)
        {
            uint version = aInvocation.Version();
            if (version != 1)
                throw new Exception(String.Format("Unexpected result - {0} - from IDvInvocation.Version", version));
            aResult = aValue + 1;
        }

        protected override void Decrement(IDvInvocation aInvocation, int aValue, out int aResult)
        {
            aResult = aValue - 1;
        }

        protected override void Toggle(IDvInvocation aInvocation, bool aValue, out bool aResult)
        {
            aResult = !aValue;
        }

        protected override void EchoString(IDvInvocation aInvocation, string aValue, out string aResult)
        {
            aResult = aValue;
        }

        protected override void EchoBinary(IDvInvocation aInvocation, byte[] aValue, out byte[] aResult)
        {
            aResult = aValue;
        }

        protected override void SetUint(IDvInvocation aInvocation, uint aValueUint)
        {
            SetPropertyVarUint(aValueUint);
        }

        protected override void GetUint(IDvInvocation aInvocation, out uint aValueUint)
        {
            aValueUint = PropertyVarUint();
        }

        protected override void SetInt(IDvInvocation aInvocation, int aValueInt)
        {
            SetPropertyVarInt(aValueInt);
        }

        protected override void GetInt(IDvInvocation aInvocation, out int aValueInt)
        {
            aValueInt = PropertyVarInt();
        }

        protected override void SetBool(IDvInvocation aInvocation, bool aValueBool)
        {
            SetPropertyVarBool(aValueBool);
        }

        protected override void GetBool(IDvInvocation aInvocation, out bool aValueBool)
        {
            aValueBool = PropertyVarBool();
        }

        protected override void SetString(IDvInvocation aInvocation, string aValueStr)
        {
            SetPropertyVarStr(aValueStr);
        }

        protected override void GetString(IDvInvocation aInvocation, out string aValueStr)
        {
            aValueStr = PropertyVarStr();
        }

        protected override void SetBinary(IDvInvocation aInvocation, byte[] aValueBin)
        {
            SetPropertyVarBin(aValueBin);
        }

        protected override void GetBinary(IDvInvocation aInvocation, out byte[] aValueBin)
        {
            aValueBin = PropertyVarBin();
        }

        protected override void SetMultiple(IDvInvocation aInvocation, uint aValueUint, int aValueInt, bool aValueBool)
        {
            PropertiesLock();
            SetPropertyVarUint(aValueUint);
            SetPropertyVarInt(aValueInt);
            SetPropertyVarBool(aValueBool);
            PropertiesUnlock();
        }
    }
    
    class DeviceBasic : IDisposable
    {
        public static string gDeviceName = "device";
        private DvDeviceStandard iDevice;
        private DvProviderOpenhomeOrgTestBasic1 iTestBasic;

        private static void RandomiseUdn(out string aUdn)
        {
            aUdn = gDeviceName;
            aUdn += "-";
            Random rnd = new Random();
            aUdn += rnd.Next();
        }

        public DeviceBasic()
        {
            RandomiseUdn(out gDeviceName);
            iDevice = new DvDeviceStandard(gDeviceName);
            iDevice.SetAttribute("Upnp.Domain", "openhome.org");
            iDevice.SetAttribute("Upnp.Type", "Test");
            iDevice.SetAttribute("Upnp.Version", "1");
            iDevice.SetAttribute("Upnp.FriendlyName", "ohNetTestDevice");
            iDevice.SetAttribute("Upnp.Manufacturer", "None");
            iDevice.SetAttribute("Upnp.ModelName", "ohNet test device");
            iTestBasic = new ProviderTestBasic(iDevice);
            iDevice.SetEnabled();
        }

        public DvDevice Device()
        {
            return iDevice;
        }
        
        public void Dispose()
        {
            iDevice.Dispose();
            iTestBasic.Dispose();
        }
    }
}

File: OpenHome/Net/Bindings/Cpp/Device/Tests/TestBasicDv.cpp
#include "TestBasicDv.h"
#include <OpenHome/Types.h>
#include <OpenHome/Net/Cpp/DvDevice.h>
#include <OpenHome/Net/Cpp/DvOpenhomeOrgTestBasic1.h>
#include <OpenHome/Private/Ascii.h>
#include <OpenHome/Private/Env.h>
#include <OpenHome/Net/Private/DviStack.h>
#include <OpenHome/Private/NetworkAdapterList.h>
#include <OpenHome/Net/Private/Globals.h>
#include <OpenHome/Private/TestFramework.h>

#include <string>

using namespace OpenHome;
using namespace OpenHome::Net;

class OpenHome::Net::ProviderTestBasic : public DvProviderOpenhomeOrgTestBasic1Cpp
{
public:
    ProviderTestBasic(DvDeviceStd& aDevice);
private:
    void Increment(IDvInvocationStd& aInvocation, uint32_t aValue, uint32_t& aResult);
    void Decrement(IDvInvocationStd& aInvocation, int32_t aValue, int32_t& aResult);
    void Toggle(IDvInvocationStd& aInvocation, bool aValue, bool& aResult);
    void EchoString(IDvInvocationStd& aInvocation, const std::string& aValue, std::string& aResult);
    void EchoBinary(IDvInvocationStd& aInvocation, const std::string& aValue, std::string& aResult);
    void SetUint(IDvInvocationStd& aInvocation, uint32_t aValueUint);
    void GetUint(IDvInvocationStd& aInvocation, uint32_t& aValueUint);
    void SetInt(IDvInvocationStd& aInvocation, int32_t aValueInt);
    void GetInt(IDvInvocationStd& aInvocation, int32_t& aValueInt);
    void SetBool(IDvInvocationStd& aInvocation, bool aValueBool);
    void GetBool(IDvInvocationStd& aInvocation, bool& aValueBool);
    void SetMultiple(IDvInvocationStd& aInvocation, uint32_t aValueUint, int32_t aValueInt, bool aValueBool);
    void SetString(IDvInvocationStd& aInvocation, const std::string& aValueStr);
    void GetString(IDvInvocationStd& aInvocation, std::string& aValueStr);
    void SetBinary(IDvInvocationStd& aInvocation, const std::string& aValueBin);
    void GetBinary(IDvInvocationStd& aInvocation, std::string& aValueBin);
};


ProviderTestBasic::ProviderTestBasic(DvDeviceStd& aDevice)
    : DvProviderOpenhomeOrgTestBasic1Cpp(aDevice)
{
    EnablePropertyVarUint();
    EnablePropertyVarInt();
    EnablePropertyVarBool();
    EnablePropertyVarStr();
    EnablePropertyVarBin();
    SetPropertyVarUint(0);
    SetPropertyVarInt(0);
    SetPropertyVarBool(false);
    std::string empty;
    SetPropertyVarStr(empty);
    SetPropertyVarBin(empty);

    EnableActionIncrement();
    EnableActionDecrement();
    EnableActionToggle();
    EnableActionEchoString();
    EnableActionEchoBinary();
    EnableActionSetUint();
    EnableActionGetUint();
    EnableActionSetInt();
    EnableActionGetInt();
    EnableActionSetBool();
    EnableActionGetBool();
    EnableActionSetMultiple();
    EnableActionSetString();
    EnableActionGetString();
    EnableActionSetBinary();
    EnableActionGetBinary();
}

void ProviderTestBasic::Increment(IDvInvocationStd& aInvocation, uint32_t aValue, uint32_t& aResult)
{
    ASSERT(aInvocation.Version() == 1);
    aResult = ++aValue;
}

void ProviderTestBasic::Decrement(IDvInvocationStd& /*aInvocation*/, int32_t aValue, int32_t& aResult)
{
    aResult = --aValue;
}

void ProviderTestBasic::Toggle(IDvInvocationStd& /*aInvocation*/, bool aValue, bool& aResult)
{
    aResult = !aValue;
}

void ProviderTestBasic::EchoString(IDvInvocationStd& /*aInvocation*/, const std::string& aValue, std::string& aResult)
{
    aResult.assign(aValue);
}

void ProviderTestBasic::EchoBinary(IDvInvocationStd& /*aInvocation*/, const std::string& aValue, std::string& aResult)
{
    aResult.assign(aValue);
}

void ProviderTestBasic::SetUint(IDvInvocationStd& /*aInvocation*/, uint32_t aValueUint)
{
    SetPropertyVarUint(aValueUint);
}

void ProviderTestBasic::GetUint(IDvInvocationStd& /*aInvocation*/, uint32_t& aValueUint)
{
    GetPropertyVarUint(aValueUint);
}

void ProviderTestBasic::SetInt(IDvInvocationStd& aInvocation, int32_t aValueInt)
{
    if (aValueInt == 12345) {
        std::string str("custom error");
        aInvocation.ReportError(801, str);
    }
    SetPropertyVarInt(aValueInt);
}

void ProviderTestBasic::GetInt(IDvInvocationStd& /*aInvocation*/, int32_t& aValueInt)
{
    GetPropertyVarInt(aValueInt);
}

void ProviderTestBasic::SetBool(IDvInvocationStd& /*aInvocation*/, bool aValueBool)
{
    SetPropertyVarBool(aValueBool);
}

void ProviderTestBasic::GetBool(IDvInvocationStd& /*aInvocation*/, bool& aValueBool)
{
    GetPropertyVarBool(aValueBool);
}

void ProviderTestBasic::SetMultiple(IDvInvocationStd& /*aInvocation*/, uint32_t aValueUint, int32_t aValueInt, bool aValueBool)
{
    PropertiesLock();
    SetPropertyVarUint(aValueUint);
    SetPropertyVarInt(aValueInt);
    SetPropertyVarBool(aValueBool);
    PropertiesUnlock();
}

void ProviderTestBasic::SetString(IDvInvocationStd& /*aInvocation*/, const std::string& aValueStr)
{
    SetPropertyVarStr(aValueStr);
}

void ProviderTestBasic::GetString(IDvInvocationStd& /*aInvocation*/, std::string& aValueStr)
{
    GetPropertyVarStr(aValueStr);
}

void ProviderTestBasic::SetBinary(IDvInvocationStd& /*aInvocation*/, const std::string& aValueBin)
{
    SetPropertyVarBin(aValueBin);
}

void ProviderTestBasic::GetBinary(IDvInvocationStd& /*aInvocation*/, std::string& aValueBin)
{
    GetPropertyVarBin(aValueBin);
}


static std::string gDeviceName("device");

static void RandomiseUdn(Environment& aEnv, std::string& aUdn)
{
    Bwh udn(aUdn.c_str());
    TestFramework::RandomiseUdn(aEnv, udn);
    aUdn.assign((const char*)udn.Ptr(), udn.Bytes());
}

DeviceBasic::DeviceBasic(Environment& aEnv, EProtocol aProtocol)
{
    RandomiseUdn(aEnv, gDeviceName);
    if (aProtocol == DeviceBasic::eProtocolNone) {
        iDevice = new DvDeviceStd(gDeviceName);
    }
    else { // eProtocolUpnp
        iDevice = new DvDeviceStdStandard(gDeviceName);
        iDevice->SetAttribute("Upnp.Domain", "openhome.org");
        iDevice->SetAttribute("Upnp.Type", "Test");
        iDevice->SetAttribute("Upnp.Version", "1");
        iDevice->SetAttribute("Upnp.FriendlyName", "ohNetTestDevice");
        iDevice->SetAttribute("Upnp.Manufacturer", "None");
        iDevice->SetAttribute("Upnp.ModelName", "ohNet test device");
    }
    iTestBasic = new ProviderTestBasic(*iDevice);
    iDevice->SetEnabled();
}

DeviceBasic::~DeviceBasic()
{
    delete iTestBasic;
    delete iDevice;
}

DvDeviceStd& DeviceBasic::Device()
{
    return *iDevice;
}
